### PR TITLE
Feature/migrate yppm_driver.f90 subroutines to yppm_core.F90

### PR DESCRIPTION
**Pull request by robgpita**
## Proposed changes

The subroutines and state variables in located yppm_driver.f90 were moved to the yppm_core.F90 file for code readability and maintainability. 

## Types of changes

What types of changes does your code introduce?
_Put an `x` in all the boxes that apply_

- [ ] Bugfix (Change which fixes an issue)
- [ ] New feature (Change which adds functionality)
- [x] Enhancement to existing functionality (non-breaking change which improves existing code or documentation)
- [ ] Breaking change (fix or feature that changes results and requires updates to test baselines)
- [ ] Documentation Update

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/NOAA-GSL/SENA-yppm/blob/develop/CONTRIBUTING.md) document
- [x] I have ensured all my changes contribute to a common purpose
- [x] I have verified my change does not add debug code
- [x] I have verified my change does not add code that is commented out
- [x] I have verified unit tests pass locally with my changes
- [x] I have verified my changes adhere to style guidelines
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

In migrating the subroutines & state variables, I removed some unnecessary code in yppm_driver.f90, notably `ONLY yppm` & `use NetCDFModule`. In the yppm_core.F90 file, I removed the lines `private` and `public yppm` as they are not necessary. The subroutines and state variables from yppm_driver.f90 were moved to yppm_core.F90.

### CODE REVIEW

**Comment by christopherwharrop-noaa** [comment]
@robgonzalezpita - Thank you for submitting this.  I'd like to get #11 merged before we tackle this one.

**Comment by robgpita** [comment]
@christopherwharrop-noaa No problem. I agree, #11 should take precedence.

**Comment by robgpita** [comment]
@christopherwharrop-noaa @LyndStringer 

I'm not entirely sure why the compare tests are failing (while the regression tests are not) with the interpolation changes.  Any Ideas? 

I'm going to give this PR a break, and move on to my other work.

**Comment by christopherwharrop-noaa** [comment]
@robgonzalezpita - I don't have time to look into it right now, but you might want to look into scoping rules for module variables vs local subroutine variables.

**Review by LyndStringer — COMMENTED** [review]
The wisdom seems to be to put the routines that need to know about the data base into yppm_core.F90 and let the driver (yppm_driver.f90) be just a high-level driver.  

I'm wondering if we might want three files out of this.  One file to just be fv3 code, that is to say, yppm_core.F90.  Then keep the driver (yppm_driver.f90) as a high level driver.  And finally, all the routines we wrote to support the driver (this would be the new file) in a separate file.  The good think about this is that anyone converting the code to a new language would only need to convert yppm_core.F90, leaving all else as is.  That is to say, if Fortran could call yppm (this new language), they need only to convert yppm_core.F90.

I'm not sure this can be done, since the new file needs to know about the database.  

What do you think?

**Comment by LyndStringer** [comment]
> I'm not entirely sure why the compare tests are failing (while the regression tests are not) with the interpolation changes. Any Ideas?

Are you saying that your ctest on Hera passes, but the CI tests do not?  

Regardless, I would diff the *.test (the expected output) files against the *.test.out (output) files, look at the differences.

PR #11 changed the *.test files (Different interpolation >> slightly different answers.).  That may be the problem.

Re: Chris' scoping suggestion.  The subroutine: print_state (and others) was inside yppm_core.F90 
because it needed detailed understanding of the data base.  That is to say, print_state was a 
subroutine in the module that declared the database, so it had scope to the data base.  When print_state is moved to another file, it losses scope of the data base unless somehow you tell it about the database.

**Comment by christopherwharrop-noaa** [comment]
I like the idea of having a clean separation from the yppm kernel code and the rest of the code that we added to support the execution of the kernel.  As you point out it would make it more clear what the "kernel" is, and what the focus should be for conversion into other approaches.  I don't know the extent to which our kernel support code would be reusable from other languages/approaches.  I think the reason the state variables did not get put into yppm_core may be for historical reasons.  The original version of the yppm had a two-layered call to yppm. I don't recall the details off hand, but the original call tree may have played a role.

**Comment by robgpita** [comment]
@christopherwharrop-noaa  I'm have no idea if the original call tree, or the two-layered call would have an impact on what I'm trying to accomplish. It is really odd to me that prior to adding the interpolation code, everything was passing, and now it does not pass. 

@LyndStringer The tests do not pass on Hera/my mac, however it is only the `compare` tests.  The CI is working fine from my understanding. The tests passed when I moved the print/write/etc. subroutines into yppm_core.F90 (without the interpolation changes). Also, I am planning on changing the 'driver' code in the Julia conversion, as the I/O will be written in Julia. I was thinking there might be a discrepancy in the ref/test/CMakeLists.txt which is pointing to different files, but I do not believe that is the case. When I move print_state back into the driver file, I'm getting the same failing test cases with `compare`.

**Comment by robgpita** [comment]
@christopherwharrop-noaa @LyndStringer Good news!!

Couldn't leave this one alone without getting to the bottom of the error. I'm not sure how it happened, but one of the write statements in the print_state subroutine was not correct. I just needed to change a 2 to a 3.

`write(*,'(A5,A20)') "TEST ", msg` should have been `write(*,'(A5,A30)') "TEST ", msg`

Something could have gone awry when I pulled the latest version of develop locally, or I made a mistake in the copy/paste.  This did provide to be a valuable experience as I gained more familiarity with the inner workings of the kernel. (CMakeLists.txt files and ref/build/bin/yppm_compare.sh)

File: ref/src/yppm_core.F90 [hunk 1]
```diff
@@ -21,11 +21,10 @@
 
 module yppm_core_mod
 
-  implicit none
-
-  private
```
Comment (christopherwharrop-noaa) [inline ref/src/yppm_core.F90:26]: It is usually best practice to make module variables and routines private by default and then use `public` to expose the ones needed outside of the module.  I do not understand your comment about these being unnecessary.  Can you please explain?

File: ref/src/yppm_core.F90 [hunk 2]
```diff
@@ -21,11 +21,10 @@
...
+  use netCDFModule
+  use interpolate
 
-  public yppm
```
Comment (christopherwharrop-noaa) [inline ref/src/yppm_core.F90:28]: This isn't needed because you removed the `private` above.  However, I'm not sure that was wise.

**Review by christopherwharrop-noaa — CHANGES_REQUESTED** [review]
I'm not following the logic of removing the `private` from the `yppm_core_mod` module and then using `public` to expose the routines needed by external clients of the module.

File: ref/src/yppm_core.F90 (hunk 1 above)
Comment (robgpita) [reply ref/src/yppm_core.F90:26]: @christopherwharrop-noaa I removed the `private` statement from this module in following the example of sw_core.F90. Now that all of the variables and routines in this module need to be exposed to yppm_driver, would we need to declare all of them as `public` here and `use` all of them by name in yppm_driver module? That would add quite a bit of redundancy, but if that is what you’d suggest, I will certainly remove this change.  I did not give this removal too much consideration, and do not have experience with the best practices/pros & cons of keeping specific variables/routines private and public.

 If it is best practice, and you would suggest keeping the module variables/routines private, I can leave the `private` statement and make futher adjustments to accommodate for this behavior as needed.

Comment (christopherwharrop-noaa) [reply ref/src/yppm_core.F90:26]: @robgonzalezpita - Thank you for explaining.  You are correct that the number of things that the driver needs would result in a long `only` clause.  However, some of the routines in this module are for internal use only, and should be private.  I would recommend we keep the routines `private` by default and use `public` to expose the ones needed outside.  And then, for now, `use` the module without the normally desired `only` clause to keep that simple.

File: ref/src/yppm_driver.f90 [hunk 3]
```diff
@@ -6,25 +6,11 @@
 program yppm_driver
 
   use OMP_LIB
-  use yppm_core_mod, ONLY: yppm
+  use yppm_core_mod
   use netCDFModule
```
Comment (christopherwharrop-noaa) [inline ref/src/yppm_driver.f90]: I think you can probably get rid of this, too?

**Review by christopherwharrop-noaa — COMMENTED** [review]
The move from the driver to the module is straightforward and I see no issues.  Test are passing.  Since all of the IO routines are now moved to `yppm_core_mod` I think we can also probably get rid of the `use` for the NetCDFModule as well.

File: ref/src/yppm_driver.f90 (hunk 3 above)
Comment (robgpita) [reply ref/src/yppm_driver.f90]: @christopherwharrop-noaa Sure can! This slipped through the cracks.

**Review by christopherwharrop-noaa — APPROVED** [review]
All of my concerns have been addressed.  @LyndStringer raises a good point about perhaps isolating our support code to another module, separate from the model code.  Let's consider that in another forum, perhaps on Slack or the issue tracking system, and talk about how we want to tackle that.

**Pull request merged**